server/assetstore.py

```diff
@@ -0,0 +1,434 @@
+            # Find or create a folder if needed
+            if 'database' not in dbinfo and parentType == 'folder':
+                folder = parent
+            else:
+                folderName = dbinfo.get('database', defaultDatabase)
+                folder = folderModel.findOne({
+                    'parentId': parent['_id'],
+                    'name': folderName,
+                    'parentCollection': parentType
+                })
+                if folder is None:
+                    folder = folderModel.createFolder(
+                        parent, folderName, parentType=parentType,
+                        creator=user)
+            # Create an item if needed
+            item = itemModel.findOne({
+                'folderId': folder['_id'],
+                'name': name
+            })
+            if item is None:
+                item = itemModel.createItem(
+                    name=name, creator=user, folder=folder)
+            # Create a file if needed
+            file = fileModel.findOne({
+                'name': name,
+                'itemId': item['_id']
+            })
+            if file is None:
+                file = fileModel.createFile(
+                    creator=user, item=item, name=name, size=0,
+                    assetstore=self.assetstore,
+                    mimeType=dbFormatList.get(preferredFormat(params.get(
+                        'format'))),
+                    saveFile=False)
+            if file.get(dbInfoKey) and not file[dbInfoKey].get('imported'):
+                raise GirderException(
+                    'A file for table %s is present but cannot be updated '
+                    'because it wasn\'t imported.' % name)
+            # Validate the limit parameter
+            try:
+                if params.get('limit') not in (None, ''):
+                    params['limit'] = int(params['limit'])
+            except ValueError:
+                raise GirderException(
+                    'limit must be empty or an integer')
+            # Set or replace the database parameters for the file
+            dbinfo['imported'] = True
+            for key in ('sort', 'fields', 'filters', 'format', 'limit'):
+                dbinfo[key] = params.get(key)
+            file[dbInfoKey] = dbinfo
+            # Validate that we can perform queries by trying to download 1
+            # record from the file.
+            #   This intentionally encodes extraParameters as JSON.  It could
+            # pass it as a python dictionary or encode it as a url query string,
+            # but another Girder plugin has expressed a preference for JSON as
+            # the de facto standard for extraParameters.
+            downloadFunc = self.downloadFile(
+                file.copy(), headers=False, extraParameters=json.dumps({
+                    'limit': 1}))
+            # Test the download without keeping it
+            [None for chunk in downloadFunc()]
```

Comment:
I think some PEP or another dictates that list comprehension syntax shouldn't be used for side-effects.